lib/web_ui/lib/src/engine/text/measurement.dart

```diff
@@ -123,5 +127,248 @@ double measureSubstring(
+  final double widthExcludingTrailingSpaces;
+
+  /// The width of the measured text, including any trailing spaces.
+  final double widthIncludingTrailingSpaces;
+
+  /// The width of trailing spaces in the measured text.
+  double get widthOfTrailingSpaces => widthIncludingTrailingSpaces - widthExcludingTrailingSpaces;
+
+  /// The total height as calculated from the font and style for this text.
+  double get height => ascent + descent;
+
+  /// The height of the measured text's bounding box.
+  double get textHeight => textAscent + textDescent;
+}
+
+/// Helper class that measures text in isolation and provides caching for better
+/// performance.
+class CachedMeasurements {
+  CachedMeasurements(this._canvasContext);
+
+  final DomCanvasRenderingContext2D _canvasContext;
+
+  /// A cache that maps: style -> text -> TextMetrics
+  final Map<String, Map<String, TextMetrics>> _cache =
+      <String, Map<String, TextMetrics>>{};
+
+  /// Measures the range of text between two line break results.
+  TextMetrics measureRange(
+    String text,
+    EngineTextStyle style,
+    LineBreakResult start,
+    LineBreakResult end,
+  ) {
+    return rawMeasure(
+      text: text,
+      style: style,
+      start: start.index,
+      endExcludingTrailingSpaces: end.indexWithoutTrailingSpaces,
+      // Trailing new lines should never be included in measurements.
+      endIncludingTrailingSpaces: end.indexWithoutTrailingNewlines,
+    );
+  }
+
+  TextMetrics measureText(String text, EngineTextStyle style) {
+    return rawMeasure(
+      text: text,
+      style: style,
+      start: 0,
+      endExcludingTrailingSpaces: text.length,
+      endIncludingTrailingSpaces: text.length,
+    );
+  }
+
+  TextMetrics measureSubstring(
+    String text,
+    EngineTextStyle style,
+    int start,
+    int end,
+  ) {
+    return rawMeasure(
+      text: text,
+      style: style,
+      start: start,
+      endExcludingTrailingSpaces: end,
+      endIncludingTrailingSpaces: end,
+    );
+  }
+
+  TextMetrics rawMeasure({
+    required String text,
+    required EngineTextStyle style,
+    required int start,
+    required int endExcludingTrailingSpaces,
+    required int endIncludingTrailingSpaces,
+  }) {
+    // 0 <= start <= endExcludingTrailingSpace <= endIncludingTrailingSpaces <= text.length.
+    assert(0 <= start);
+    assert(start <= endExcludingTrailingSpaces);
+    assert(endExcludingTrailingSpaces <= endIncludingTrailingSpaces);
+    assert(endIncludingTrailingSpaces <= text.length);
+
+    // Empty range.
+    if (start == endIncludingTrailingSpaces) {
+      return TextMetrics.zero;
+    }
+
+    final String cssFontString = style.cssFontString;
+    print('font: "$cssFontString"');
+
+    final String textIncludingTrailingSpaces =
+        _extractSubstring(text, start, endIncludingTrailingSpaces);
+
+    final Map<String, TextMetrics> fontCache =
+        _cache.putIfAbsent(cssFontString, () => <String, TextMetrics>{});
```

Comment:
We might want to try avoiding allocating closures in the middle of measuring text, if possible.